sstables/generation_type.hh

```diff
@@ -57,9 +64,31 @@ public:
             ec == std::errc() && ptr == s.data() + s.size()) {
             return generation_type(int_value);
         } else {
-            throw std::invalid_argument(fmt::format("invalid UUID: {}", s));
+            static const boost::regex pattern("([0-9a-z]{4})_([0-9a-z]{4})_([0-9a-z]{5})([0-9a-z]{13})");
+            boost::smatch match;
+            if (!boost::regex_match(s, match, pattern)) {
+                throw std::invalid_argument(fmt::format("invalid UUID: {}", s));
+            }
+            utils::UUID_gen::decimicroseconds timestamp = {};
+            auto decode_base36 = [](const std::string& s) {
+                std::size_t pos{};
+                auto n = std::stoull(s, &pos, 36);
+                if (pos != s.size()) {
+                    throw std::invalid_argument(fmt::format("invalid part in UUID: {}", s));
+                }
+                return n;
+            };
+            timestamp += std::chrono::days{decode_base36(match[1])};
+            timestamp += std::chrono::seconds{decode_base36(match[2])};
+            timestamp += ::utils::UUID_gen::decimicroseconds{decode_base36(match[3])};
+            int64_t lsb = decode_base36(match[4]);
+            return generation_type{utils::UUID_gen::get_time_UUID_raw(timestamp, lsb)};
```

Comment:
I feel we should go for a similar implementation of base36 as the one we have in utils/base64.hh. no regex. can be tested independently in an unit test. your test below is good, but I feel like we should stay away from regex and go for a utils/base36.hh. see: https://github.com/scylladb/scylladb/pull/12398/commits/2014c19f0fff7fd9b46a91732b14986afdb31cd4#diff-1ced7176c7c62252ef00444af0603fa1e72d9e109db66ea598a8584bbdadb33cR20

Reply:
could you help me understand why we should "stay away from regex" here? and why a dedicated `base36.hh` source file is better? IMHO, the base36 encode/decode is very trivial here. it is implemented using `std::stoull()` and `std::to_chars()`.  do we really need a header file for the wrappers around these two functions?

Reply:
I think a local regex based parser is fine here. This is a custom format, not something generic, and over-generalization will just result in complicated code but without any benefit from reuse.

Reply:
Makes sense.